docs/source/howto/data.rst

```diff
@@ -248,11 +247,247 @@ The node class has various methods to set these attributes, such as :py:`~aiida.
+Since the ``CalcJobNode`` *created* the ``Int`` node, it is an *incoming* node from the ``Int`` node's perspective.
+
+In the context of our query, we are building a *path* consisting of *vertices* (i.e. the entities we query for) connected by *edges* defined by the relationships between them.
+A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.
+
+.. _how-to:data:find:projections:
+
+Projections
+-----------
+
+When no *projection* is specified, the default behavior of the ``QueryBuilder`` is to project the entities corresponding to the final vertex of the query path.
+However, in many cases we are not interested in the entities themselves, but rather their PK, UUID, *attributes* or some other piece of information stored by the entity.
+This can be achieved using the ``project`` keyword argument:
+
+.. code-block:: python
+
+    qb = QueryBuilder()
+    qb.append(CalcJobNode, tag='calcjob')
+    qb.append(Int, with_incoming='calcjob', project='id')
+
+In the above example, executing the query returns all PK's of the ``Int`` nodes which are outputs of all ``CalcJobNode``'s in the database.
+However, you can project more than one piece of information for one vertex by using a list:
+
+.. code-block:: python
+
+    qb = QueryBuilder()
+    qb.append(CalcJobNode, tag='calcjob')
+    qb.append(Int, with_incoming='calcjob', project=['id', 'attributes.value'])
+
+Moreover, you can project information for multiple vertices along the query path:
+
+.. code-block:: python
+
+    qb = QueryBuilder()
+    qb.append(CalcJobNode, tag='calcjob', project='uuid)
+    qb.append(Int, with_incoming='calcjob', project=['id', 'attributes.value'])
+
+Asking only for the properties that you are interested in can result in much faster queries.
+If you want the Aiida-ORM instance, add ``'*'`` to your list of projections:
+
+.. code-block:: python
+
+    qb = QueryBuilder()
+    qb.append(CalcJobNode, tag='calcjob', project='uuid')
+    qb.append(Int, with_incoming='calcjob', project=['id', '*', 'attributes.value'])
+
+All projections must start with one of the *columns* of the entities in the database.
+Examples of columns we have encountered so far are ``id``, ``uuid`` and ``attributes``.
+If the column is a dictionary, you can expand the dictionary values using a dot notation, as we have done in the previous example to obtain the ``attributes.value``.
+This can be used to project the values of nested dictionaries as well.
+
+.. note::
 
-`#3996`_
+    Be aware that, for consistency, ``QueryBuilder.all()`` / ``iterall()`` always returns a list of lists, even if you project on one entity.
+    A convenient boolean keyword argument for the ``all()`` method is ``flat``, which returns the query as a flat list if set to ``True``.
```

Comment:
should we point out that this really often only makes sense if you are projecting a single property of a single entity?

Reply:
I've adjusted the first sentence to highlight this some more, and added "in this case" to the second:

```
.. note::

    Be aware that, for consistency, ``QueryBuilder.all()`` / ``iterall()`` always returns a list of lists, even if you only project one property of a single entity.
    A convenient boolean keyword argument for the ``all()`` method in this case is ``flat``, which returns the query as a flat list if set to ``True``.
```